### PR TITLE
Added draw to MagickImage, DrawingWand class and a few drawables

### PR DESCRIPTION
**Pull request by Peeterush**
There's still a couple of things to resolve:

`DrawableFillOpacity` only works when the fill color has been set with an alpha between 1-254 already. If an alpha of 255 has been set, `DrawableFillOpacity` does not seem to be able to overwrite it. If an alpha of 0 has been set before, `DrawableFillOpacity` does actually work, but then its percentage is used inverted. Is this all expected behaviour?

I can't get `DrawableFont` to work, causing `DrawableText` to error `NonconformingDrawingPrimitiveDefinition`. It seems setting the font on the `DrawingSettings` object that is used to create a `DrawingWand` instance works, but assigning a font to an already created `DrawingWand` instance does not. Any ideas what's going wrong?

Furthermore, I don't know how to write tests for `MagickImage.draw()`, `DrawingSettings` and `DrawingWand`.
A separate `MagickImage.draw()` would probably perform a copy of a test to any of the drawables' tests, making it redundant.
A `DrawingSettings` test would need access to `MagickImage._settings`, which is private.
`DrawingWand` is something to be used by `MagickImage`, not something to be exposed in the API.
They're all used by the tests for the drawables themselves of course, but since something like `MagickImage` also tests its constructor, I figured you probably also want separate tests for these as well. So any ideas on this?

I'd love to hear your thoughts!

### CODE REVIEW

**Comment by Peeterush** [comment]
@dlemstra any input on this?

**Comment by dlemstra** [comment]
I haven't had time to check this. My focus has been elsewhere. I might be able to take a better look on Friday.

**Comment by dlemstra** [comment]
I don't really understand your `DrawableFillOpacity` question but alpha and opacity are inverted. This means that alpha 0 is 255 opacity.

And I figured out what is causing the `NonconformingDrawingPrimitiveDefinition` error. I just pushed a patch that adds `relatedErrors` to the MagickError and it turns out that somewhere inside ImageMagick you get `MagickError: UnableToReadFont `helvetica' @ error/annotate.c/RenderFreetype/1474`. Will need to check why this is happening.

**Comment by dlemstra** [comment]
Found the problem. You are calling `_DrawingSettings_Font_Set` instead of `_DrawingWand_Font`. When I change that no exception is thrown.

I also don't think we need the `fontFamily` because we always need to use a font file.

In Magick.NET I don't have separate tests for `DrawingWand` so testing this through `image.draw` should be sufficient.

**Comment by dlemstra** [comment]
Also just added a `Magick._getFontFileName` that you could use.

**Comment by dlemstra** [comment]
Did you close this PR by accident @Peeterush?

**Comment by Peeterush** [comment]
Yes that was not intentional. I think I will have to make a new one, as I don't seem to be able to reopen it.

**Comment by Peeterush** [comment]
Thanks for all your input! Hopefully I can complete this now.

> I also don't think we need the `fontFamily` because we always need to use a font file.

Is it not possible for one file to contain multiple styles, weights or stretches? If that is possible, there still needs to be a function to access those. But tell me if you rather see it gone, I'll remove it then.

> I don't really understand your DrawableFillOpacity question but alpha and opacity are inverted. This means that alpha 0 is 255 opacity.

Yes but that is not what confuses me. Consider a black image, modified by the following:
```
image.draw([
    new DrawableFillColor(new MagickColor(255, 0, 0)),
    new DrawableFillOpacity(new Percentage(10)),
    new DrawableColor(0, 0, PaintMethod.Floodfill)
]);
```
I would expect the image to be filled with RGBA 255,0,0,26 but instead the source turns RGBA 255,0,0,255. `DrawableFillOpacity` doesn't do anything here. To get the expected output I would first have to set an alpha on `DrawableFillColor` like this:
```
image.draw([
    new DrawableFillColor(new MagickColor(255, 0, 0, 128)),
    new DrawableFillOpacity(new Percentage(10)),
    new DrawableColor(0, 0, PaintMethod.Floodfill)
]);
```
Now the image comes out as expected. This 128 value could be any other value, but not 255 or 0. A value of 255 would be the same as not providing an alpha. But 0 is an interesting case. The following snippet will yield the same result as the previous snippet (mind the percentage change):
```
image.draw([
    new DrawableFillColor(new MagickColor(255, 0, 0, 0)),
    new DrawableFillOpacity(new Percentage(90)),
    new DrawableColor(0, 0, PaintMethod.Floodfill)
]);
```
With a pre-set alpha of 0, the percentage of `DrawableFillOpacity` is flipped.

So the two things I am confused about:
- I need to set a random alpha before even being allowed to use `DrawableFillOpacity`.
- When an alpha of 0 has been set, the percentage given to `DrawableFillOpacity` is flipped.

**Comment by dlemstra** [comment]
> Is it not possible for one file to contain multiple styles, weights or stretches? If that is possible, there still needs to be a function to access those. But tell me if you rather see it gone, I'll remove it then.

I don't think we can support that because only fonts from a file are supported. Not everything that is possible in the .NET library is possible with this library. Better to remove it for now.

And thanks for all the info about `DrawableFillOpacity`. Will need to investigate why that happens.

File: src/drawables/drawable-color.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Copyright Dirk Lemstra https://github.com/dlemstra/magick-wasm.
+// Licensed under the Apache License, Version 2.0.
+
+import { PaintMethod } from "../paint-method";
```
Comment (dlemstra) [inline src/drawables/drawable-color.ts:4]: Can you sort these alphabetically? And also in other files?

File: src/drawables/drawable-color.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// Copyright Dirk Lemstra https://github.com/dlemstra/magick-wasm.
+// Licensed under the Apache License, Version 2.0.
+
+import { PaintMethod } from "../paint-method";
+import { IDrawable } from "./drawable";
+import { IDrawingWand } from "./drawing-wand";
+
+export class DrawableColor implements IDrawable {
+    _x: number;
```
Comment (dlemstra) [inline src/drawables/drawable-color.ts:9]: Can you mark these fields `private readonly`? And also in the other files?

File: src/drawables/drawable-font.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+// Licensed under the Apache License, Version 2.0.
+
+import { IDrawable } from "./drawable";
+import { IDrawingWand } from "./drawing-wand";
+import { Magick } from "../magick";
+
+export class DrawableFont implements IDrawable {
+    _family: string;
+
+    constructor(family: string) {
```
Comment (dlemstra) [inline src/drawables/drawable-font.ts:11]: Maybe we should change this to `font` instead of `family`?

File: src/drawables/drawing-wand.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
+// Copyright Dirk Lemstra https://github.com/dlemstra/magick-wasm.
+// Licensed under the Apache License, Version 2.0.
+
+import { ImageMagick } from '../image-magick';
+import { Exception } from '../internal/exception/exception';
+import { NativeInstance } from '../internal/native-instance';
+import { _withString } from '../internal/native/string';
```
Comment (dlemstra) [inline src/drawables/drawing-wand.ts:7]: I put these at the bottom.

File: src/drawables/drawing-wand.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    static _create(image: IMagickImage, settings: MagickSettings): DrawingWand {
+        return new DrawingWand(image, settings);
+    }
+
+    draw(drawables: IDrawable[]): void {
+        drawables.forEach(drawable => {
+            drawable.draw(this)
+        });
+
+        return Exception.usePointer(exception => {
```
Comment (dlemstra) [inline src/drawables/drawing-wand.ts:46]: Do we need `return` when the return type is `void`?

File: src/settings/drawing-settings.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+    fontPointsize?: number;
+
+    strokeColor?: MagickColor;
+
+    strokeWidth?: number;
+
+    /** @internal */
+    static _create(settings: MagickSettings): DrawingSettings {
+        const instace = new DrawingSettings();
```
Comment (dlemstra) [inline src/settings/drawing-settings.ts:23]: Small typo.

File: src/settings/native-drawing-settings.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        super(instance, disposeMethod);
+
+        if (settings.fillColor !== undefined) {
+            settings.fillColor._use(valuePtr => {
+                ImageMagick._api._DrawingSettings_FillColor_Set(this._instance, valuePtr);
+            });
+        }
+
+        if (settings.font !== undefined) {
+            const fileName = `/fonts/${settings.font}`;
```
Comment (dlemstra) [inline src/settings/native-drawing-settings.ts:23]: You could use `Magick._getFontFileName` here?

File: src/settings/native-magick-settings.ts [hunk 8]
```diff
@@ -24,7 +24,7 @@ export class NativeMagickSettings extends NativeInstance {
             ImageMagick._api._MagickSettings_SetQuality(this._instance, settings._quality);
 
         if (settings.backgroundColor !== undefined) {
-            settings.backgroundColor ._use((ptr) => {
+            settings.backgroundColor._use((ptr) => {
```
Comment (dlemstra) [inline src/settings/native-magick-settings.ts:27]: Thanks :)

**Review by dlemstra — COMMENTED** [review]
Thanks for your help. Most of the feedback is coding style but I don't mind doing them myself.

**Comment by Peeterush** [comment]
All your remarks should be resolved!

I think most of the things you requested are things ESLint can point out, and even automatically fix, given a stricter configuration. Would you consider that?

**Comment by dlemstra** [comment]
Probably a better idea to do this after this pull request has been merged but it would be nice if we could autofix this with a stricter configuration.

**Review by dlemstra — APPROVED** [review]
Thanks for helping me out with this!

**Comment by dlemstra** [comment]
I have found the issue with `DrawableFillOpacity` and pushed a fix to the ImageMagick repository. Thanks for helping us find this issue.

**Comment by Peeterush** [comment]
Great! I suppose the 128 in `drawable-fill-opacity.spec.ts` can be removed then?

**Comment by dlemstra** [comment]
I will take a look at this after I upgrade Magick.Native here.

**Comment by Peeterush** [comment]
I checked the fix, and indeed `DrawableFillOpacity` now works without previously setting an alpha on the fill color. But the second part of my confusion has not been fixed: if you first set a fill color with alpha 0, then use `DrawableFillOpacity`, the percentage is inverted.
```
image.draw([
    new DrawableFillColor(new MagickColor(255, 0, 0)),
    new DrawableFillOpacity(new Percentage(10)),
    new DrawableColor(0, 0, PaintMethod.Floodfill)
]);

// is equal to

image.draw([
    new DrawableFillColor(new MagickColor(255, 0, 0, 128)),
    new DrawableFillOpacity(new Percentage(10)),
    new DrawableColor(0, 0, PaintMethod.Floodfill)
]);

// and equal to

image.draw([
    new DrawableFillColor(new MagickColor(255, 0, 0, 255)),
    new DrawableFillOpacity(new Percentage(10)),
    new DrawableColor(0, 0, PaintMethod.Floodfill)
]);

// but is not equal to

image.draw([
    new DrawableFillColor(new MagickColor(255, 0, 0, 0)),
    new DrawableFillOpacity(new Percentage(10)),
    new DrawableColor(0, 0, PaintMethod.Floodfill)
]);
```

**Comment by dlemstra** [comment]
We just made another patch to resolve this. Will probably update ImageMagick here some time next week.

**Comment by Peeterush** [comment]
Great! Thank you for all your work.

**Pull request merged**